File: protocol-generator/python/src/entity_core/peer/typedefs.py
"""The V7 §9.5 core type-registry floor (53 types).

CLEAN-ROOM: the type definitions below are an in-code MODEL rendered through the
peer's own S2 codec; the served entity's content_hash is computed by our own
encoder over the model, NOT ingested from oracle bytes.  The shapes follow the
V7 §9.5 floor + the cross-peer render-from-model ruling (single source of truth
in code; omit-empty so the canonical ECF map is byte-stable regardless of field
declaration order — the codec re-sorts map keys length-then-lex).

A *core* peer publishes EXACTLY the §9.5 floor (core + operational + the
type-system bootstrap).  Extension vocabularies (compute/*, content/*, …) are
NOT pre-published by a core peer (refined G4 / F17).
"""

from __future__ import annotations

from typing import Any

from .model import Entity


# ── field-spec builders (omit-empty) ──────────────────────────────────────────
def _fref(t: str) -> dict:
    return {"type_ref": t}


def _opt(s: dict) -> dict:
    s = dict(s)
    s["optional"] = True
    return s


def _sized(n: int, s: dict) -> dict:
    s = dict(s)
    s["byte_size"] = n
    return s


def _farray(elem: dict) -> dict:
    return {"array_of": elem}


def _fmap_of(key_type: str, val: dict) -> dict:
    d: dict[str, Any] = {"map_of": val}
    if key_type:
        d["key_type"] = key_type
    return d


def _funion(*variants: dict) -> dict:
    return {"union_of": list(variants)}


# reused nested specs
_sp_string = _fref("primitive/string")
_sp_any = _fref("primitive/any")
_sp_hash = _fref("system/hash")
_sp_core_entity = _fref("core/entity")
_sp_tree_path = _fref("system/tree/path")
_sp_grant_entry = _fref("system/capability/grant-entry")
_sp_multi_granter = _fref("system/capability/multi-granter")
_sp_field_spec = _fref("system/type/field-spec")
_sp_op_spec = _fref("system/handler/operation-spec")
_sp_listing_entry = _fref("system/tree/listing-entry")
_sp_type = _fref("system/type")
_sp_type_name = _fref("system/type/name")


# ── type-def builder ──────────────────────────────────────────────────────────
def _type_def(
    name: str,
    *,
    extends: str | None = None,
    fields: list[tuple[str, dict]] | None = None,
    layout: list[str] | None = None,
) -> dict:
    """Render a system/type data map (omit-empty)."""
    d: dict[str, Any] = {"name": name}
    if extends is not None:
        d["extends"] = extends
    if fields:
        d["fields"] = {fname: fspec for fname, fspec in fields}
    if layout:
        d["layout"] = list(layout)
    return d


def _core_type_defs() -> list[dict]:
    """The V7 §9.5 core type floor (53 definitions) as system/type data maps."""
    return [
        # primitives (8)
        _type_def("primitive/any"),
        _type_def("primitive/bool"),
        _type_def("primitive/bytes"),
        _type_def("primitive/float"),
        _type_def("primitive/int"),
        _type_def("primitive/null"),
        _type_def("primitive/string"),
        _type_def("primitive/uint"),
        # structural roots + envelopes (5)
        _type_def("entity", fields=[
            ("type", _fref("primitive/string")),
            ("data", _fref("primitive/any")),
        ]),
        _type_def("core/entity", fields=[
            ("type", _fref("primitive/string")),
            ("data", _fref("primitive/any")),
            ("content_hash", _fref("system/hash")),
        ]),
        _type_def("core/envelope", fields=[
            ("root", _fref("core/entity")),
            ("included", _opt(_fmap_of("system/hash", _sp_core_entity))),
        ]),
        _type_def("system/envelope", extends="core/envelope"),
        _type_def("system/protocol/envelope", extends="core/envelope"),
        # identity / hash / signature (4)
        _type_def("system/hash", extends="primitive/bytes", fields=[
            ("format_code", _sized(1, _fref("primitive/uint"))),
            ("digest", _fref("primitive/bytes")),
        ], layout=["format_code", "digest"]),
        _type_def("system/peer", fields=[
            ("key_type", _fref("primitive/string")),
            ("peer_id", _fref("system/peer-id")),
            ("public_key", _fref("primitive/bytes")),
        ]),
        _type_def("system/peer-id", extends="primitive/string"),
        _type_def("system/signature", fields=[
            ("algorithm", _fref("primitive/string")),
            ("signature", _fref("primitive/bytes")),
            ("signer", _fref("system/hash")),
            ("target", _fref("system/hash")),
        ]),
        # protocol surface (6)
        _type_def("system/protocol/connect/authenticate", fields=[
            ("key_type", _fref("primitive/string")),
            ("nonce", _fref("primitive/bytes")),
            ("peer_id", _fref("system/peer-id")),
            ("public_key", _fref("primitive/bytes")),
        ]),
        _type_def("system/protocol/connect/hello", fields=[
            ("protocols", _farray(_sp_string)),
            ("nonce", _fref("primitive/bytes")),
            ("peer_id", _fref("system/peer-id")),
            ("timestamp", _fref("primitive/uint")),
            ("compression", _opt(_farray(_sp_string))),
            ("encryption", _opt(_farray(_sp_string))),
            ("hash_formats", _opt(_farray(_sp_string))),
            ("key_types", _opt(_farray(_sp_string))),
        ]),
        _type_def("system/protocol/error", fields=[
            ("code", _fref("primitive/string")),
            ("message", _opt(_fref("primitive/string"))),
            ("rejected_marker", _opt(_fref("system/hash"))),
        ]),
        _type_def("system/protocol/execute", fields=[
            ("operation", _fref("primitive/string")),
            ("params", _fref("core/entity")),
            ("request_id", _fref("primitive/string")),
            ("uri", _fref("system/tree/path")),
            ("author", _opt(_fref("system/hash"))),
            ("bounds", _opt(_fref("system/bounds"))),
            ("capability", _opt(_fref("system/hash"))),
            ("deliver_to", _opt(_fref("system/delivery-spec"))),
            ("deliver_token", _opt(_fref("system/hash"))),
            ("durability_request", _opt(_fref("system/durability-request"))),
            ("resource", _opt(_fref("system/protocol/resource-target"))),
        ]),
        _type_def("system/protocol/execute/response", fields=[
            ("request_id", _fref("primitive/string")),
            ("result", _fref("core/entity")),
            ("status", _fref("primitive/uint")),
            ("durability", _opt(_fref("system/durability-result"))),
        ]),
        _type_def("system/protocol/resource-target", fields=[
            ("targets", _farray(_sp_tree_path)),
            ("exclude", _opt(_farray(_sp_tree_path))),
        ]),
        # capability (12)
        _type_def("system/capability/grant", fields=[
            ("token", _fref("system/hash")),
        ]),
        _type_def("system/capability/grant-entry", fields=[
            ("handlers", _fref("system/capability/path-scope")),
            ("operations", _fref("system/capability/id-scope")),
            ("resources", _fref("system/capability/path-scope")),
            ("allowances", _opt(_fmap_of("", _sp_any))),
            ("constraints", _opt(_fmap_of("", _sp_any))),
            ("peers", _opt(_fref("system/capability/id-scope"))),
        ]),
        _type_def("system/capability/id-scope", fields=[
            ("include", _farray(_sp_string)),
            ("exclude", _opt(_farray(_sp_string))),
        ]),
        _type_def("system/capability/path-scope", fields=[
            ("include", _farray(_sp_tree_path)),
            ("exclude", _opt(_farray(_sp_tree_path))),
        ]),
        _type_def("system/capability/request", fields=[
            ("grants", _farray(_sp_grant_entry)),
            ("ttl_ms", _opt(_fref("primitive/uint"))),
        ]),
        _type_def("system/capability/revocation", fields=[
            ("token", _fref("system/hash")),
            ("revoked_at", _fref("primitive/uint")),
            ("reason", _opt(_fref("primitive/string"))),
        ]),
        _type_def("system/capability/revoke-request", fields=[
            ("token", _fref("system/hash")),
            ("reason", _opt(_fref("primitive/string"))),
        ]),
        _type_def("system/capability/delegate-request", fields=[
            ("grants", _farray(_sp_grant_entry)),
            ("parent", _fref("system/hash")),
            ("ttl_ms", _opt(_fref("primitive/uint"))),
        ]),
        _type_def("system/capability/delegation-caveats", fields=[
            ("max_delegation_depth", _opt(_fref("primitive/uint"))),
            ("max_delegation_ttl", _opt(_fref("primitive/uint"))),
            ("no_delegation", _opt(_fref("primitive/bool"))),
        ]),
        _type_def("system/capability/policy-entry", fields=[
            ("grants", _farray(_sp_grant_entry)),
            ("peer_pattern", _fref("primitive/string")),
            ("notes", _opt(_fref("primitive/string"))),
            ("ttl_ms", _opt(_fref("primitive/uint"))),
        ]),
        _type_def("system/capability/token", fields=[
            ("created_at", _fref("primitive/uint")),
            ("grantee", _fref("system/hash")),
            ("granter", _funion(_sp_hash, _sp_multi_granter)),
            ("grants", _farray(_sp_grant_entry)),
            ("delegation_caveats", _opt(_fref("system/capability/delegation-caveats"))),
            ("expires_at", _opt(_fref("primitive/uint"))),
            ("not_before", _opt(_fref("primitive/uint"))),
            ("parent", _opt(_fref("system/hash"))),
            ("resource_limits", _opt(_fref("system/resource-limits"))),
        ]),
        _type_def("system/capability/multi-granter", fields=[
            ("signers", _farray(_sp_hash)),
            ("threshold", _fref("primitive/uint")),
        ]),
        # handler machinery (6)
        _type_def("system/handler", fields=[
            ("interface", _fref("system/tree/path")),
            ("expression_path", _opt(_fref("system/tree/path"))),
            ("internal_scope", _opt(_farray(_sp_grant_entry))),
            ("max_scope", _opt(_farray(_sp_grant_entry))),
        ]),
        _type_def("system/handler/interface", fields=[
            ("name", _fref("primitive/string")),
            ("operations", _fmap_of("", _sp_op_spec)),
            ("pattern", _fref("system/tree/path")),
        ]),
        _type_def("system/handler/manifest", extends="system/handler/interface", fields=[
            ("name", _fref("primitive/string")),
            ("operations", _fmap_of("", _sp_op_spec)),
            ("pattern", _fref("system/tree/path")),
            ("expression_path", _opt(_fref("system/tree/path"))),
            ("internal_scope", _opt(_farray(_sp_grant_entry))),
            ("max_scope", _opt(_farray(_sp_grant_entry))),
        ]),
        _type_def("system/handler/operation-spec", fields=[
            ("input_type", _opt(_fref("system/type/name"))),
            ("output_type", _opt(_fref("system/type/name"))),
        ]),
        _type_def("system/handler/register-request", fields=[
            ("manifest", _fref("system/handler/manifest")),
            ("requested_scope", _opt(_farray(_sp_grant_entry))),
            ("types", _opt(_fmap_of("", _sp_type))),
        ]),
        _type_def("system/handler/register-result", fields=[
            ("grant", _fref("system/capability/token")),
            ("pattern", _fref("system/tree/path")),
        ]),
        # tree (5)
        _type_def("system/tree/get-request", fields=[
            ("limit", _opt(_fref("primitive/uint"))),
            ("mode", _opt(_fref("primitive/string"))),
            ("offset", _opt(_fref("primitive/uint"))),
            ("tree_id", _opt(_fref("primitive/string"))),
        ]),
        _type_def("system/tree/put-request", fields=[
            ("entity", _opt(_fref("core/entity"))),
            ("expected_hash", _opt(_fref("system/hash"))),
            ("tree_id", _opt(_fref("primitive/string"))),
        ]),
        _type_def("system/tree/listing", fields=[
            ("count", _fref("primitive/uint")),
            ("entries", _fmap_of("", _sp_listing_entry)),
            ("offset", _fref("primitive/uint")),
            ("path", _fref("system/tree/path")),
            ("next_page", _opt(_fref("system/hash"))),
        ]),
        _type_def("system/tree/listing-entry", fields=[
            ("has_children", _fref("primitive/bool")),
            ("hash", _opt(_fref("system/hash"))),
        ]),
        _type_def("system/tree/path", extends="primitive/string"),
        # type-system bootstrap (3)
        _type_def("system/type", fields=[
            ("name", _fref("system/type/name")),
            ("extends", _opt(_fref("system/type/name"))),
            ("fields", _opt(_fmap_of("", _sp_field_spec))),
            ("layout", _opt(_farray(_sp_string))),
            ("type_args", _opt(_fmap_of("", _sp_type_name))),
            ("type_params", _opt(_farray(_sp_string))),
        ]),
        _type_def("system/type/field-spec", fields=[
            ("type_ref", _opt(_fref("system/type/name"))),
            ("optional", _opt(_fref("primitive/bool"))),
            ("array_of", _opt(_fref("system/type/field-spec"))),
            ("map_of", _opt(_fref("system/type/field-spec"))),
            ("union_of", _opt(_farray(_sp_field_spec))),
            ("key_type", _opt(_fref("system/type/name"))),
            ("byte_size", _opt(_fref("primitive/uint"))),
            ("type_param", _opt(_fref("primitive/string"))),
            ("type_args", _opt(_fmap_of("", _sp_type_name))),
            ("default", _opt(_fref("primitive/any"))),
            ("constraints", _opt(_farray(_sp_core_entity))),
        ]),
        _type_def("system/type/name", extends="primitive/string"),
        # operational (4)
        _type_def("system/bounds", fields=[
            ("budget", _opt(_fref("primitive/uint"))),
            ("cascade_depth", _opt(_fref("primitive/uint"))),
            ("chain_id", _opt(_fref("primitive/string"))),
            ("parent_chain_id", _opt(_fref("primitive/string"))),
            ("ttl", _opt(_fref("primitive/uint"))),
            ("visited", _opt(_farray(_sp_tree_path))),
        ]),
        _type_def("system/resource-limits", fields=[
            ("max_budget", _opt(_fref("primitive/uint"))),
            ("max_ttl", _opt(_fref("primitive/uint"))),
            ("max_visited_length", _opt(_fref("primitive/uint"))),
        ]),
        _type_def("system/delivery-spec", fields=[
            ("operation", _fref("primitive/string")),
            ("uri", _fref("system/tree/path")),
        ]),
        _type_def("system/deletion-marker"),
    ]


def core_type_entities() -> list[tuple[str, Entity]]:
    """Materialize each §9.5 floor type as a (name, system/type Entity) pair."""
    out = []
    for td in _core_type_defs():
        out.append((td["name"], Entity.make("system/type", td)))
    return out

File: protocol-generator/go/src/peer/typedefs.go
package peer

// typedefs.go — the V7 §9.5 core type-registry floor (53 types).
//
// CLEAN-ROOM: the type definitions below are an in-code MODEL rendered through
// the peer's own S2 codec; the served entity's content_hash is computed by our
// own encoder over the model, not ingested from oracle bytes. The shapes follow
// the V7 §9.5 floor + the cross-peer render-from-model ruling (single source of
// truth in code, omit-empty so the canonical ECF map is byte-stable regardless
// of field declaration order — the codec re-sorts map keys length-then-lex).
//
// A *core* peer publishes EXACTLY the §9.5 floor (core + operational + the
// type-system bootstrap). Extension vocabularies (compute/*, content/*, …) are
// NOT pre-published by a core peer (refined G4 / F17): the oracle's type_system
// category matches the floor as a hard FAIL gate and WARNs (matched-if-present)
// on the non-floor types it also probes.

import "github.com/entity-core/entity-core-protocol-go/internal/cbor"

// fspec is a field spec inside a typeDef (the system/type/field-spec shape).
// Exactly one structural carrier is normally set: a typeRef, an arrayOf, a
// mapOf, or a unionOf. Rendered omit-empty into the field-spec ECF map.
type fspec struct {
	typeRef  string
	optional bool
	arrayOf  *fspec
	mapOf    *fspec
	unionOf  []fspec
	keyType  string
	byteSize uint64
	hasSize  bool
}

// fref → a type_ref to a named type.
func fref(t string) fspec { return fspec{typeRef: t} }

// opt marks a field-spec optional.
func opt(s fspec) fspec { s.optional = true; return s }

// sized adds a byte_size carrier.
func sized(n uint64, s fspec) fspec { s.byteSize = n; s.hasSize = true; return s }

// farray → an array_of the given element spec.
func farray(elem fspec) fspec { e := elem; return fspec{arrayOf: &e} }

// fmapOf → a map_of the given value spec, with an optional key_type ("" = none).
func fmapOf(keyType string, val fspec) fspec {
	v := val
	return fspec{mapOf: &v, keyType: keyType}
}

// funion → a union_of the given variant specs.
func funion(vs ...fspec) fspec { return fspec{unionOf: vs} }

// data renders a field-spec to its ECF data map (omit-empty). The codec sorts
// keys canonically; only the present key/value set affects the bytes.
func (s fspec) data() cbor.Value {
	var p []cbor.Pair
	if s.typeRef != "" {
		p = append(p, cbor.Entry("type_ref", cbor.Text(s.typeRef)))
	}
	if s.optional {
		p = append(p, cbor.Entry("optional", cbor.Bool(true)))
	}
	if s.arrayOf != nil {
		p = append(p, cbor.Entry("array_of", s.arrayOf.data()))
	}
	if s.mapOf != nil {
		p = append(p, cbor.Entry("map_of", s.mapOf.data()))
	}
	if s.unionOf != nil {
		variants := make([]cbor.Value, len(s.unionOf))
		for i, v := range s.unionOf {
			variants[i] = v.data()
		}
		p = append(p, cbor.Entry("union_of", valList(variants...)))
	}
	if s.keyType != "" {
		p = append(p, cbor.Entry("key_type", cbor.Text(s.keyType)))
	}
	if s.hasSize {
		p = append(p, cbor.Entry("byte_size", cbor.Uint(s.byteSize)))
	}
	return cbor.NewMap(p...)
}

// field is one named field of a typeDef (declaration order preserved within the
// fields sub-map; the codec re-sorts it canonically on encode).
type field struct {
	name string
	spec fspec
}

func f(name string, spec fspec) field { return field{name, spec} }

// typeDef is a core type definition (the system/type entity data).
type typeDef struct {
	name    string
	extends string
	fields  []field
	layout  []string
}

func def(name string) typeDef { return typeDef{name: name} }

func (t typeDef) withExtends(e string) typeDef { t.extends = e; return t }
func (t typeDef) withFields(fs ...field) typeDef {
	t.fields = fs
	return t
}
func (t typeDef) withLayout(ls ...string) typeDef { t.layout = ls; return t }

// data renders the system/type data map (omit-empty).
func (t typeDef) data() cbor.Value {
	var p []cbor.Pair
	p = append(p, cbor.Entry("name", cbor.Text(t.name)))
	if t.extends != "" {
		p = append(p, cbor.Entry("extends", cbor.Text(t.extends)))
	}
	if len(t.fields) > 0 {
		fp := make([]cbor.Pair, len(t.fields))
		for i, fl := range t.fields {
			fp[i] = cbor.Pair{Key: cbor.Text(fl.name), Val: fl.spec.data()}
		}
		p = append(p, cbor.Entry("fields", cbor.NewMap(fp...)))
	}
	if len(t.layout) > 0 {
		p = append(p, cbor.Entry("layout", strList(t.layout...)))
	}
	return cbor.NewMap(p...)
}

// entity materializes the type as a system/type entity (content_hash computed by
// our own encoder over the model).
func (t typeDef) entity() Entity { return mustEntity("system/type", t.data()) }

// reused nested specs.
var (
	spString       = fref("primitive/string")
	spAny          = fref("primitive/any")
	spHash         = fref("system/hash")
	spCoreEntity   = fref("core/entity")
	spTreePath     = fref("system/tree/path")
	spGrantEntry   = fref("system/capability/grant-entry")
	spMultiGranter = fref("system/capability/multi-granter")
	spFieldSpec    = fref("system/type/field-spec")
	spOpSpec       = fref("system/handler/operation-spec")
	spListingEntry = fref("system/tree/listing-entry")
	spType         = fref("system/type")
	spTypeName     = fref("system/type/name")
)

// coreTypeDefs is the V7 §9.5 core type floor (53 definitions).
func coreTypeDefs() []typeDef {
	return []typeDef{
		// primitives (8)
		def("primitive/any"),
		def("primitive/bool"),
		def("primitive/bytes"),
		def("primitive/float"),
		def("primitive/int"),
		def("primitive/null"),
		def("primitive/string"),
		def("primitive/uint"),

		// structural roots + envelopes (5)
		def("entity").withFields(
			f("type", fref("primitive/string")),
			f("data", fref("primitive/any")),
		),
		def("core/entity").withFields(
			f("type", fref("primitive/string")),
			f("data", fref("primitive/any")),
			f("content_hash", fref("system/hash")),
		),
		def("core/envelope").withFields(
			f("root", fref("core/entity")),
			f("included", opt(fmapOf("system/hash", spCoreEntity))),
		),
		def("system/envelope").withExtends("core/envelope"),
		def("system/protocol/envelope").withExtends("core/envelope"),

		// identity / hash / signature (4)
		def("system/hash").withExtends("primitive/bytes").withFields(
			f("format_code", sized(1, fref("primitive/uint"))),
			f("digest", fref("primitive/bytes")),
		).withLayout("format_code", "digest"),
		def("system/peer").withFields(
			f("key_type", fref("primitive/string")),
			f("peer_id", fref("system/peer-id")),
			f("public_key", fref("primitive/bytes")),
		),
		def("system/peer-id").withExtends("primitive/string"),
		def("system/signature").withFields(
			f("algorithm", fref("primitive/string")),
			f("signature", fref("primitive/bytes")),
			f("signer", fref("system/hash")),
			f("target", fref("system/hash")),
		),

		// protocol surface (6)
		def("system/protocol/connect/authenticate").withFields(
			f("key_type", fref("primitive/string")),
			f("nonce", fref("primitive/bytes")),
			f("peer_id", fref("system/peer-id")),
			f("public_key", fref("primitive/bytes")),
		),
		def("system/protocol/connect/hello").withFields(
			f("protocols", farray(spString)),
			f("nonce", fref("primitive/bytes")),
			f("peer_id", fref("system/peer-id")),
			f("timestamp", fref("primitive/uint")),
			f("compression", opt(farray(spString))),
			f("encryption", opt(farray(spString))),
			f("hash_formats", opt(farray(spString))),
			f("key_types", opt(farray(spString))),
		),
		def("system/protocol/error").withFields(
			f("code", fref("primitive/string")),
			f("message", opt(fref("primitive/string"))),
			f("rejected_marker", opt(fref("system/hash"))),
		),
		def("system/protocol/execute").withFields(
			f("operation", fref("primitive/string")),
			f("params", fref("core/entity")),
			f("request_id", fref("primitive/string")),
			f("uri", fref("system/tree/path")),
			f("author", opt(fref("system/hash"))),
			f("bounds", opt(fref("system/bounds"))),
			f("capability", opt(fref("system/hash"))),
			f("deliver_to", opt(fref("system/delivery-spec"))),
			f("deliver_token", opt(fref("system/hash"))),
			f("durability_request", opt(fref("system/durability-request"))),
			f("resource", opt(fref("system/protocol/resource-target"))),
		),
		def("system/protocol/execute/response").withFields(
			f("request_id", fref("primitive/string")),
			f("result", fref("core/entity")),
			f("status", fref("primitive/uint")),
			f("durability", opt(fref("system/durability-result"))),
		),
		def("system/protocol/resource-target").withFields(
			f("targets", farray(spTreePath)),
			f("exclude", opt(farray(spTreePath))),
		),

		// capability (12)
		def("system/capability/grant").withFields(
			f("token", fref("system/hash")),
		),
		def("system/capability/grant-entry").withFields(
			f("handlers", fref("system/capability/path-scope")),
			f("operations", fref("system/capability/id-scope")),
			f("resources", fref("system/capability/path-scope")),
			f("allowances", opt(fmapOf("", spAny))),
			f("constraints", opt(fmapOf("", spAny))),
			f("peers", opt(fref("system/capability/id-scope"))),
		),
		def("system/capability/id-scope").withFields(
			f("include", farray(spString)),
			f("exclude", opt(farray(spString))),
		),
		def("system/capability/path-scope").withFields(
			f("include", farray(spTreePath)),
			f("exclude", opt(farray(spTreePath))),
		),
		def("system/capability/request").withFields(
			f("grants", farray(spGrantEntry)),
			f("ttl_ms", opt(fref("primitive/uint"))),
		),
		def("system/capability/revocation").withFields(
			f("token", fref("system/hash")),
			f("revoked_at", fref("primitive/uint")),
			f("reason", opt(fref("primitive/string"))),
		),
		def("system/capability/revoke-request").withFields(
			f("token", fref("system/hash")),
			f("reason", opt(fref("primitive/string"))),
		),
		def("system/capability/delegate-request").withFields(
			f("grants", farray(spGrantEntry)),
			f("parent", fref("system/hash")),
			f("ttl_ms", opt(fref("primitive/uint"))),
		),
		def("system/capability/delegation-caveats").withFields(
			f("max_delegation_depth", opt(fref("primitive/uint"))),
			f("max_delegation_ttl", opt(fref("primitive/uint"))),
			f("no_delegation", opt(fref("primitive/bool"))),
		),
		def("system/capability/policy-entry").withFields(
			f("grants", farray(spGrantEntry)),
			f("peer_pattern", fref("primitive/string")),
			f("notes", opt(fref("primitive/string"))),
			f("ttl_ms", opt(fref("primitive/uint"))),
		),
		def("system/capability/token").withFields(
			f("created_at", fref("primitive/uint")),
			f("grantee", fref("system/hash")),
			f("granter", funion(spHash, spMultiGranter)),
			f("grants", farray(spGrantEntry)),
			f("delegation_caveats", opt(fref("system/capability/delegation-caveats"))),
			f("expires_at", opt(fref("primitive/uint"))),
			f("not_before", opt(fref("primitive/uint"))),
			f("parent", opt(fref("system/hash"))),
			f("resource_limits", opt(fref("system/resource-limits"))),
		),
		def("system/capability/multi-granter").withFields(
			f("signers", farray(spHash)),
			f("threshold", fref("primitive/uint")),
		),

		// handler machinery (6)
		def("system/handler").withFields(
			f("interface", fref("system/tree/path")),
			f("expression_path", opt(fref("system/tree/path"))),
			f("internal_scope", opt(farray(spGrantEntry))),
			f("max_scope", opt(farray(spGrantEntry))),
		),
		def("system/handler/interface").withFields(
			f("name", fref("primitive/string")),
			f("operations", fmapOf("", spOpSpec)),
			f("pattern", fref("system/tree/path")),
		),
		def("system/handler/manifest").withExtends("system/handler/interface").withFields(
			f("name", fref("primitive/string")),
			f("operations", fmapOf("", spOpSpec)),
			f("pattern", fref("system/tree/path")),
			f("expression_path", opt(fref("system/tree/path"))),
			f("internal_scope", opt(farray(spGrantEntry))),
			f("max_scope", opt(farray(spGrantEntry))),
		),
		def("system/handler/operation-spec").withFields(
			f("input_type", opt(fref("system/type/name"))),
			f("output_type", opt(fref("system/type/name"))),
		),
		def("system/handler/register-request").withFields(
			f("manifest", fref("system/handler/manifest")),
			f("requested_scope", opt(farray(spGrantEntry))),
			f("types", opt(fmapOf("", spType))),
		),
		def("system/handler/register-result").withFields(
			f("grant", fref("system/capability/token")),
			f("pattern", fref("system/tree/path")),
		),

		// tree (5)
		def("system/tree/get-request").withFields(
			f("limit", opt(fref("primitive/uint"))),
			f("mode", opt(fref("primitive/string"))),
			f("offset", opt(fref("primitive/uint"))),
			f("tree_id", opt(fref("primitive/string"))),
		),
		def("system/tree/put-request").withFields(
			f("entity", opt(fref("core/entity"))),
			f("expected_hash", opt(fref("system/hash"))),
			f("tree_id", opt(fref("primitive/string"))),
		),
		def("system/tree/listing").withFields(
			f("count", fref("primitive/uint")),
			f("entries", fmapOf("", spListingEntry)),
			f("offset", fref("primitive/uint")),
			f("path", fref("system/tree/path")),
			f("next_page", opt(fref("system/hash"))),
		),
		def("system/tree/listing-entry").withFields(
			f("has_children", fref("primitive/bool")),
			f("hash", opt(fref("system/hash"))),
		),
		def("system/tree/path").withExtends("primitive/string"),

		// type-system bootstrap (3)
		def("system/type").withFields(
			f("name", fref("system/type/name")),
			f("extends", opt(fref("system/type/name"))),
			f("fields", opt(fmapOf("", spFieldSpec))),
			f("layout", opt(farray(spString))),
			f("type_args", opt(fmapOf("", spTypeName))),
			f("type_params", opt(farray(spString))),
		),
		def("system/type/field-spec").withFields(
			f("type_ref", opt(fref("system/type/name"))),
			f("optional", opt(fref("primitive/bool"))),
			f("array_of", opt(fref("system/type/field-spec"))),
			f("map_of", opt(fref("system/type/field-spec"))),
			f("union_of", opt(farray(spFieldSpec))),
			f("key_type", opt(fref("system/type/name"))),
			f("byte_size", opt(fref("primitive/uint"))),
			f("type_param", opt(fref("primitive/string"))),
			f("type_args", opt(fmapOf("", spTypeName))),
			f("default", opt(fref("primitive/any"))),
			f("constraints", opt(farray(spCoreEntity))),
		),
		def("system/type/name").withExtends("primitive/string"),

		// operational (4)
		def("system/bounds").withFields(
			f("budget", opt(fref("primitive/uint"))),
			f("cascade_depth", opt(fref("primitive/uint"))),
			f("chain_id", opt(fref("primitive/string"))),
			f("parent_chain_id", opt(fref("primitive/string"))),
			f("ttl", opt(fref("primitive/uint"))),
			f("visited", opt(farray(spTreePath))),
		),
		def("system/resource-limits").withFields(
			f("max_budget", opt(fref("primitive/uint"))),
			f("max_ttl", opt(fref("primitive/uint"))),
			f("max_visited_length", opt(fref("primitive/uint"))),
		),
		def("system/delivery-spec").withFields(
			f("operation", fref("primitive/string")),
			f("uri", fref("system/tree/path")),
		),
		def("system/deletion-marker"),
	}
}

// publishCoreTypes writes the §9.5 floor at /{peer}/system/type/{name}.
func (p *Peer) publishCoreTypes() {
	for _, td := range coreTypeDefs() {
		p.store.Bind("/"+p.localPeer+"/system/type/"+td.name, td.entity())
	}
}
